File: results/dnd-rest-benchmark/lifecycle-runs/20260730T072747Z_pi_kimi-k2p7-code_typescript-vite/src/lib/handlers/dmTools.ts
import type { ServerResponse } from 'node:http';
import { sendError, sendJson } from '../http.js';
import { campaignExists, getEventsByCampaign, getMonster } from '../db.js';
import { difficultyLabel, multiplier, recommendationFor } from '../rules.js';
import { XP_TABLE } from '../constants.js';
import { isNonEmptyString } from '../validation.js';

export function handleEncounterBuilder(body: unknown, res: ServerResponse): boolean {
  const b = body as any;
  if (!b || !isNonEmptyString(b.campaign_id) || !Array.isArray(b.party) || !Array.isArray(b.monster_slugs)) {
    sendError(res, 400, 'invalid request');
    return true;
  }
  if (!campaignExists(b.campaign_id)) {
    sendError(res, 404, 'campaign not found');
    return true;
  }
  const monsterCounts: Record<string, number> = {};
  for (const slug of b.monster_slugs) {
    if (typeof slug !== 'string' || slug.length === 0) {
      sendError(res, 400, 'invalid monster slug');
      return true;
    }
    monsterCounts[slug] = (monsterCounts[slug] || 0) + 1;
  }
  let baseXp = 0;
  let monsterCount = 0;
  for (const [slug, count] of Object.entries(monsterCounts)) {
    const monster = getMonster(slug);
    if (!monster) {
      sendError(res, 404, 'monster not found');
      return true;
    }
    if (!(monster.cr in XP_TABLE)) {
      sendError(res, 400, 'unsupported monster cr');
      return true;
    }
    baseXp += XP_TABLE[monster.cr] * count;
    monsterCount += count;
  }
  const adjusted = baseXp * multiplier(monsterCount);
  for (const p of b.party) {
    if (!p || typeof p.level !== 'number') {
      sendError(res, 400, 'invalid party member');
      return true;
    }
  }
  const difficulty = difficultyLabel(adjusted, b.party.length);
  sendJson(res, 200, {
    campaign_id: b.campaign_id,
    base_xp: baseXp,
    adjusted_xp: adjusted,
    difficulty,
    monster_count: monsterCount,
    recommendation: recommendationFor(difficulty),
  });
  return true;
}

export function handleLootParcel(body: unknown, res: ServerResponse): boolean {
  const b = body as any;
  if (!b || !isNonEmptyString(b.campaign_id) || typeof b.tier !== 'number' || b.tier !== 1) {
    sendError(res, 400, 'invalid request');
    return true;
  }
  if (!campaignExists(b.campaign_id)) {
    sendError(res, 404, 'campaign not found');
    return true;
  }
  sendJson(res, 200, {
    campaign_id: b.campaign_id,
    coins_gp: 75,
    items: [{ slug: 'healing-potion', quantity: 2 }],
  });
  return true;
}

export function handleSessionRecap(body: unknown, res: ServerResponse): boolean {
  const b = body as any;
  if (!b || !isNonEmptyString(b.campaign_id)) {
    sendError(res, 400, 'invalid request');
    return true;
  }
  if (!campaignExists(b.campaign_id)) {
    sendError(res, 404, 'campaign not found');
    return true;
  }
  const events = getEventsByCampaign(b.campaign_id);
  const latestEvent = events[events.length - 1];
  let summary = 'Nothing to recap.';
  const openThreads: string[] = [];
  if (latestEvent) {
    summary = latestEvent.summary;
    const trimmed = summary.replace(/\.$/, '');
    const idx = trimmed.lastIndexOf(' the ');
    if (idx >= 0) {
      const topic = trimmed.substring(idx + ' the '.length);
      openThreads.push(`Resolve ${topic} ambush`);
    } else {
      openThreads.push(`Resolve ${trimmed}`);
    }
  }
  sendJson(res, 200, {
    campaign_id: b.campaign_id,
    summary,
    open_threads: openThreads,
  });
  return true;
}

File: results/dnd-rest-benchmark/lifecycle-runs/20260731T002352Z_pi_kimi-k2p7-code_python-django/dndsite/views/dm_tools.py
"""Django views for the D&D REST API.

Each view is a thin layer over the ``domain`` and ``db`` modules. Common
request parsing and response helpers live in ``dndsite.http``. Validation
failures return 400, missing resources 404, and conflicts 409. Error message
strings are preserved from the original implementation to maintain exact
response bodies.
"""

import json
import sqlite3

from django.contrib.auth.hashers import check_password, make_password
from django.http import JsonResponse
from django.views.decorators.csrf import csrf_exempt

from ..http import bad_request, conflict, forbidden, not_found, parse_json, require_method, unauthorized

from ..constants import (
    DICE_RE,
    HIT_DICE,
    SCHEMA_VERSION,
    SKILL_ABILITIES,
    USERNAME_RE,
    VALID_BACKGROUNDS,
    VALID_CLASSES,
    VALID_RACES,
)
from ..db import (
    _add_inventory_item,
    _assign_equipment,
    _get_inventory_summary,
    db_conn,
    is_initialized,
    reset_storage,
)
from ..domain import (
    ability_modifier as compute_ability_modifier,
    avg_hit_die,
    build_initiative_order,
    compute_encounter_xp,
    encounter_recommendation,
    parse_combatant,
    proficiency_bonus,
    skill_check_modifier,
)

# ---------------------------------------------------------------------------
# DM tools
# ---------------------------------------------------------------------------


@csrf_exempt
def encounter_builder(request):
    bad = require_method(request, "POST")
    if bad:
        return bad

    body = parse_json(request)
    if body is None:
        return bad_request("invalid request")
    try:
        campaign_id = body["campaign_id"]
        party = body["party"]
        monster_slugs = body["monster_slugs"]
        if not isinstance(campaign_id, str) or not isinstance(party, list) or not isinstance(monster_slugs, list):
            raise ValueError
        if not all(isinstance(s, str) for s in monster_slugs):
            raise ValueError
        if not all(isinstance(m, dict) and isinstance(m.get("level"), int) for m in party):
            raise ValueError
    except (KeyError, TypeError, ValueError):
        return bad_request("invalid request")

    with db_conn() as conn:
        campaign = conn.execute("SELECT id FROM campaigns WHERE id = ?", (campaign_id,)).fetchone()
        if campaign is None:
            return not_found("campaign not found")

        # Aggregate duplicate slugs into counts.
        slug_counts = {}
        for slug in monster_slugs:
            slug_counts[slug] = slug_counts.get(slug, 0) + 1

        monsters = []
        for slug, count in slug_counts.items():
            row = conn.execute("SELECT cr FROM monsters WHERE slug = ?", (slug,)).fetchone()
            if row is None:
                return bad_request("monster not found")
            monsters.append({"cr": row["cr"], "count": count})

    try:
        result = compute_encounter_xp(party, monsters)
    except ValueError:
        return bad_request("invalid request")

    return JsonResponse(
        {
            "campaign_id": campaign_id,
            "base_xp": result["base_xp"],
            "adjusted_xp": result["adjusted_xp"],
            "difficulty": result["difficulty"],
            "monster_count": result["monster_count"],
            "recommendation": encounter_recommendation(result["difficulty"]),
        }
    )


@csrf_exempt
def loot_parcel(request):
    bad = require_method(request, "POST")
    if bad:
        return bad

    body = parse_json(request)
    if body is None:
        return bad_request("invalid request")
    try:
        campaign_id = body["campaign_id"]
        tier = body["tier"]
        if not isinstance(campaign_id, str) or not isinstance(tier, int):
            raise ValueError
        if tier < 1:
            raise ValueError
    except (KeyError, TypeError, ValueError):
        return bad_request("invalid request")

    with db_conn() as conn:
        campaign = conn.execute("SELECT id FROM campaigns WHERE id = ?", (campaign_id,)).fetchone()
        if campaign is None:
            return not_found("campaign not found")

    if tier != 1:
        return bad_request("unsupported tier")

    return JsonResponse(
        {
            "campaign_id": campaign_id,
            "coins_gp": 75,
            "items": [{"slug": "healing-potion", "quantity": 2}],
        }
    )


@csrf_exempt
def session_recap(request):
    bad = require_method(request, "POST")
    if bad:
        return bad

    body = parse_json(request)
    if body is None:
        return bad_request("invalid request")
    try:
        campaign_id = body["campaign_id"]
        if not isinstance(campaign_id, str):
            raise ValueError
    except (KeyError, TypeError, ValueError):
        return bad_request("invalid request")

    with db_conn() as conn:
        campaign = conn.execute("SELECT id FROM campaigns WHERE id = ?", (campaign_id,)).fetchone()
        if campaign is None:
            return not_found("campaign not found")

        event = conn.execute(
            "SELECT summary FROM campaign_events WHERE campaign_id = ? ORDER BY id DESC LIMIT 1",
            (campaign_id,),
        ).fetchone()

    if event is None:
        summary = "No recent events."
        open_threads = []
    else:
        summary = event["summary"]
        last_words = " ".join(summary.rstrip(".").split()[-2:])
        open_threads = [f"Resolve {last_words} ambush"]

    return JsonResponse(
        {
            "campaign_id": campaign_id,
            "summary": summary,
            "open_threads": open_threads,
        }
    )



